index.html

```diff
@@ -1377,8 +1377,9 @@ <h3>Types</h3>
         </p>
 
         <p>
-All <a>credentials</a>, <a>presentations</a>, and encapsulated objects MUST
-specify, or be associated with, additional more narrow <a>types</a> (like
+If <a>credentials</a>, <a>presentations</a>, and encapsulated objects have
+properties beyond those described in their respective base types, they MUST
+specify, or be associated with, additional more narrow <a>types</a> (such as
 <code>UniversityDegreeCredential</code>, for example) so software systems can
 process this additional information.
```

Comment:
Hmmm....
```
If software systems are expected to process information in properties of
<a>credentials</a>, <a>presentations</a>, or encapsulated objects beyond the
properties described in their respective base types, such <a>credentials</a>,
<a>presentations</a>, or encapsulated objects MUST specify, or be associated
with, additional more narrow <a>types</a> in which those additional properties
are described (for example, <code>UniversityDegreeCredential</code>). Lacking
this additional type association or specification, such additional properties
MAY be ignored by systems which process these <a>credentials</a>,
<a>presentations</a>, or encapsulated objects.
```
?

Reply:
My change makes clear that additional types are only a MUST if there are additional properties not currently defined by a type, thereby softening the requirement in cases where an additional type is nonsensical.
Your recommendation softens the type requirement further than I like, essentially changing it from a MUST to a MAY. It also mentions expected processing, which I am not comfortable adding to the spec.

Reply:
I do not believe I softened anything, though my language appears to have made some things clearer, and exposed an existing softness with which you apparently disagree.

We already have (and you preserved) this great big handwave of "or be associated with" within the MUST, which removes the need to explicitly declare a sub-type within the object at hand, making it that explicit declaration a MAY.

To my knowledge, we do not require that every Validator must be able to understand every property within every VC/VP, nor that every Validator must associate every extra-property-bearing VC/VP with a sub-type of its declared base type -- only effectively that such association be possible in the abstract (i.e., that such a sub-type must be described *somewhere* [and implicitly that such sub-type must be understood by the Issuer and by *some* but not necessarily *all* potential Validators] but not that such description be accessible by any-and-all potential Validators).  Even there, I have not understood this to be a requirement for *every* property but for every *material* property, which materiality can only be determined by the Validator.

Essentially, we've said that _Validators that care_ about the extra property must be able to associate the VC/VP with a sub-type description which describes that property. In other words, Validators must be able to access a property description _for every property with which they're concerned,_ via a description of an associated type for every VC/VP they're processing *which type association **MAY** be declared within the VC/VP, or which subtype association **MAY** be inferred based on the presence of some property/ies within the VC/VP which are not described by the declared base type.*

Reply:
I appreciate that we have a tension between:
- verifiers that do not particularly care about additional properties and will accept any VC regardless of their additional properties and regardless of any constraints the issuer may have placed on the VC, and 
- verifiers that do care and want to understand every property in a VC otherwise they will not accept it.
I think that the former should be outside the scope of the standard, otherwise what is the point of standardising features that the former will happily ignore.
I think the standard should cater for the later, the well behaved eco-system citizens. Thus I do not like @TallTed's addtional text as this is pandering to the former gung-ho citizens.
So to address @TallTed's point "we do not require that every Validator must be able to understand every property within every VC/VP" I think the standard MUST provide mechanisms so that every verifier is able to understand every property in every VC/VP. Then those verifiers that do not care can ignore some properties, but those verifiers that do care can understand every property.

Reply:
Whatever we decide here, we should remember that there may be verifiers that understand particular properties (e.g. "ageOver") that do not care or know about particular VC subtypes.

Reply:
> pandering to the former gung-ho citizens

There is an unfortunate tone here. Frankly, to my eyes, it changes the rest of your comment from a technical argument to a religious one.

To the technical --

* A Verifier whose policy is that it must understand all properties in every VC/VP is free to reject any VC/VP that lacks the explicit sub-type declarations which would deliver that understanding for otherwise unrecognized properties.

* A Verifier who only cares about the properties of the base type is free to accept any VC/VP of that base type, regardless of any additional properties and/or sub-type declaration.

* A Verifier who cares about some specific sub-type is free to require either the explicit declaration _(which remains optional!)_ or the specific additional property/ies (which allow the VC/VP to be associated with that sub-type without any explicit declaration) of that sub-type.

* An Issuer who knows about a sub-type may include the additional property/ies required by that sub-type, with or without explicit declaration of that sub-type, which might have a URI which is only dereferenceable by agents doing so from within a particular IP address range or DNS (sub)domain.

An example...  MyCollege issues a VC about me.  MyCollege includes only the base type declaration (Credential), with properties for the degree name (Bachelor of Science in Biology), honors (Summa Cum Laude), sub-discipline (Reptiles), sub-sub-discipline (Snakes).  The obvious extra properties are implicitly associated with MyCollege's BachelorDegreeDetailed VC type -- which it does not describe outside the university network -- and the BachelorDegree is not sub-typed at all -- even to say it is a DegreeCredential.

The current interpretation suggested by @brentzundel and @David-Chadwick  (which does matches neither the old nor new wording, as I read things) is that this VC is invalid, because it does not explicitly declare its sub-type -- even though it can "be associated with" that sub-type by both its Issuer and any Verifier who knows about MyCollege's sub-type.

Reply:
@dlongley > Whatever we decide here, we should remember that there may be verifiers that understand particular properties (e.g. "ageOver") that do not care or know about particular VC subtypes.

Without knowing the context of a property (which is indicated by the type of the VC), then in general it will make little sense to a verifier to simply know that an isolated property is present. What is this the ageOver property the age over of? Is it the subject, a relative of the subject, or a property of the subject, or a property of some other property etc. (Thought experiment. Replace all additional property names (except ageOver and the base properties that all VCs must contain) by XXXX in any arbitrary VC, and vary the position of ageOver between these XXXX properties, and then see of you can determine what ageOver really means. Clearly you cannot. This is why the type property is essential to understanding the VC.

Reply:
@David-Chadwick,

> Without knowing the context of a property (which is indicated by the type of the VC)...

The context is indicated by the `@context` property. Types can be a shortcut for this, but they needn't be. You can define "top level" properties that apply no matter where they are and then use the fact that it's a graph model to understand which subject (in the subject/predicate/object model) to which they apply.

> ...then in general it will make little sense to a verifier to simply know that an isolated property is present. What is this the ageOver property the age over of? Is it the subject, a relative of the subject, or a property of the subject, or a property of some other property etc.

Again, it's the property of whichever node in the graph upon which it appears. This allows verifiers to only care about "ageOver" and look for it on `credentialSubject` (for example) -- without having to care whether it comes from, whether that be a driver's license, a student ID, a birth certificate, etc. The verifier only has to care about "ageOver" and that the VC comes from an issuer they trust. The graph model tells them whether or not it applies to the entity for which they have some requirement.

> (Thought experiment. Replace all additional property names (except ageOver and the base properties that all VCs must contain) by XXXX in any arbitrary VC, and vary the position of ageOver between these XXXX properties, and then see of you can determine what ageOver really means. Clearly you cannot. This is why the type property is essential to understanding the VC.

I feel like there's still some misunderstanding here either about how the graph model works or what the use case is as, if I understand you correctly, this is a trivial exercise:

```js
{
    "...",
    "credentialSubject": {
      "id": "ex:abc",
      "ageOver": "18",
      "XXXX": {
        "id": "ex:def",
        "ageOver": "21",
        "XXXX": {
           "id": "ex:ghi",
           "ageOver": "18",
           "XXXX": /* and so on */
        }
      }
    }
  }
}
```

This will produce triples:

<ex:abc> <ageOver> "18" .
<ex:def> <ageOver> "21" .
<ex:ghi> <ageOver> "18" .

Which mean:

The entity identified by "ex:abc" is over the age of 18.
The entity identified by "ex:def" is over the age of 18.
The entity identified by "ex:ghi" is over the age of 18.

Which tells me exactly which entities the "ageOver" property applies to and I don't have to care in the slightest what the various "XXXX" properties (or types) are. If I can find a statement that says "ex:abc" (or whatever ID I care about) is over the requisite age, then I'm good, that's all I care about. Again, this is one good reason why we adopted a graph model in the first place.

Reply:
Great explanation. But note that the "id" is optional, so the verifier might only be aware that some unknown entity has ageOver. Eg. credentialSubject has XXX has ageOver 18. Is XXX the subject's car? or daughter? or criminalRecord?

Reply:
@David-Chadwick,

> But note that the "id" is optional, so the verifier might only be aware that some unknown entity has ageOver. Eg. credentialSubject has XXX has ageOver 18. Is XXX the subject's car? or daughter? or criminalRecord?

Sure. But that's why the verifier asks for a VC with an "ageOver" claim that applies to whomever (i.e, has a matching ID) is authenticating with it. The verifier doesn't have to care what type of VC it is, only that the ID + "ageOver" (or some sensible way to link them) is present. Similarly, if you don't have an "id" that can sensibly be linked to whomever is authenticating it doesn't matter if you have type information or know other properties either.

Reply:
If the Verifier cannot associate `ageOver` with the Subject with which they are concerned, the Verifier is free to (and clearly should) refuse the requested action.  The `"id"` is optional *in the VC Data Model,* but it _may be required by the Verifier's Policy._

Consider the real-world age verification policies of alcohol sellers. Many will accept IDs issued in their local region (e.g., a local driver's license) or by their nation (e.g., a domestic passport), but not IDs issued in *any* region (e.g., a driver's license from a distant state, or a foreign passport) -- because it's relatively easy to know how to check validity of the local documents (basic formatting, watermarks, holographic imagery, etc.), but quite difficult to know how to check validity of every such document from around the world (does Russia include a hologram on their passports? does Moscow include a watermark in the lamination of their driver's license?)...

In other words --

* The VC Data Model is (and should be) flexible.  

* VC Verifier policies need not (and generally should probably not) be (as) flexible.